### ISSUE TITLE
[ALM Accelerator - BUG]: Export pipeline fails with an npm error

### ISSUE THREAD

**Issue by mdpc-nnit** [microsoft/coe-starter-kit#1794]
### Describe the issue

During export, it fails in the **update deployment settings**  for some solutions step with the following error:

```
2022-01-19T14:20:05.1339233Z Updating files: 100% (5712/5712), done.
2022-01-19T14:20:38.4570677Z npm WARN deprecated @types/open@6.2.1: This is a stub types definition. open provides its own type definitions, so you do not need this installed.
2022-01-19T14:20:38.5866237Z npm WARN deprecated querystring@0.2.1: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
2022-01-19T14:20:43.3569367Z npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
2022-01-19T14:21:08.0130222Z 
2022-01-19T14:21:08.0131792Z added 784 packages, and audited 785 packages in 49s
2022-01-19T14:21:08.0132263Z 
2022-01-19T14:21:08.0132679Z 95 packages are looking for funding
2022-01-19T14:21:08.0133113Z   run `npm fund` for details
2022-01-19T14:21:08.0170197Z 
2022-01-19T14:21:08.0171473Z 7 vulnerabilities (1 moderate, 6 high)
2022-01-19T14:21:08.0171746Z 
2022-01-19T14:21:08.0172213Z To address issues that do not require attention, run:
2022-01-19T14:21:08.0172679Z   npm audit fix
2022-01-19T14:21:08.0172893Z 
2022-01-19T14:21:08.0173347Z To address all issues (including breaking changes), run:
2022-01-19T14:21:08.0173816Z   npm audit fix --force
2022-01-19T14:21:08.0174051Z 
2022-01-19T14:21:08.0174418Z Run `npm audit` for details.
2022-01-19T14:21:09.3306224Z 
2022-01-19T14:21:09.3307399Z > coe@0.0.1 build
2022-01-19T14:21:09.3307858Z > tsc
2022-01-19T14:21:09.3308084Z 
2022-01-19T14:21:20.9722814Z 
2022-01-19T14:21:20.9724149Z added 1 package, and audited 3 packages in 747ms
2022-01-19T14:21:20.9734163Z 
2022-01-19T14:21:20.9734742Z found 0 vulnerabilities
2022-01-19T14:21:22.2732982Z error: option '-s, --settings <namevalues>' argument missing
2022-01-19T14:21:22.3929409Z 
2022-01-19T14:21:22.3929881Z 
2022-01-19T14:21:22.5754355Z ##[error]PowerShell exited with code '1'.
```

It seems to fail only on solutions where pipelines aren't already set up.

### Expected Behavior

Export succeeds, and deployment pipelines are set up(?)

### What component are you experiencing the issue with?

ALM Accelerator Pipelines

### What solution version are you using?

January release 2022

**Comment by mdpc-nnit** [comment]
nvm - just managed to fail an already onboarded app.

**Comment by Ballzer0** [comment]
We are also experiencing this issue, but im confused now. I managed to initially commit one of my solutions once before it stopped working.

Then i tried another solution and it seems to work all the time on that one.

**Comment by mikefactorial** [comment]
@mdpc-nnit and @Ballzer0 can you verify if the deployment profile associated with the solution has deployment steps associated with it? Does look like we aren't handling the scenario if there are none correctly, but not sure if that's the issue or not. Could also be an upgrade issue if you didn't visit / revisit the deployment configuration screens. The error is pretty specific that the -s parameter is not specified and the pipelines use the deployment configuration and associated steps to build that.

**Comment by Ballzer0** [comment]
> @mdpc-nnit and @Ballzer0 can you verify if the deployment profile associated with the solution has deployment steps associated with it? Does look like we aren't handling the scenario if there are none correctly, but not sure if that's the issue or not. Could also be an upgrade issue if you didn't visit / revisit the deployment configuration screens. The error is pretty specific that the -s parameter is not specified and the pipelines use the deployment configuration and associated steps to build that.

@mikefactorial  the solution im trying to export only contains tables, no apps, no flows and no connection references. its all empty, this is as expected.
![image](https://user-images.githubusercontent.com/28560729/150301903-36a874ac-acf8-43c6-a837-8ab09d613af1.png)

**Comment by mdpc-nnit** [comment]
@mikefactorial I am still very confused about the new functionality, so please tell me where to find deployment steps.

**Comment by Ballzer0** [comment]
> @mikefactorial I am still very confused about the new functionality, so please tell me where to find deployment steps.

i second this

**Comment by devkeydet** [comment]
Sorry you are experiencing this with the latest release.  We'll do our best to prioritize helping unblock you.  Once we do, we could use your help with feedback on how to make our docs, videos, and the overall experience better for others.
https://docs.microsoft.com/en-us/power-platform/guidance/coe/almacceleratorpowerplatform-components

We love actionable feedback.

**Comment by mikefactorial** [comment]
@mdpc-nnit this article describes some of the administrative functions introduced in the latest release. Specifically setting up profiles / user settings. https://docs.microsoft.com/en-us/power-platform/guidance/coe/setup-almacceleratorpowerplatform-deployment-profiles. I'm going to put together an accompanying video for this doc as well in the next couple of days and will follow up with that.

Assuming you created the profile from the app it should prepopulate the deployment steps. With validation, test, prod. Deployment steps provide a mapping between a target deployment environment N:1 and a profile 1:N. You can see this configuration in the ALM Accelerator for Power Platform Administration app. 
![image](https://user-images.githubusercontent.com/42348035/150349897-d29651a5-58cf-43af-8e85-de4c75f8b752.png)

I'm working on duplicating the issue you are seeing and should be able to report back more later today. Thanks again for the feedback.

**Comment by mikefactorial** [comment]
I haven't been able to replicate this, but I did notice that there were some issues in the logic of the update deployment settings template that could lead to the issue. I've tried to guard against in the template in this PR https://github.com/microsoft/coe-alm-accelerator-templates/pull/107. 

My assumption is the following.

The pipeline generation in the update deployment settings code assumed that if there was a mismatch between the number of deployment pipelines configured in Azure DevOps and the number of deployment steps configured for the deployment profile in Dataverse it would attempt to create the deployment pipelines. NOTE: This logic DEPENDS on the naming convention of the deployment pipelines in Azure DevOps being deploy-[step name e.g. validation, test, prod]-SolutionName. The logic is a little bit brittle as a result. We weren't previously guarding against there being 

1. 0 Steps
2. More Pipelines than Steps

I've updated the logic to only run the pipeline generation logic if there are specifically fewer existing pipelines than steps.

If you wouldn't mind replacing the update-deployment-settings.yml you deployed previously with the file from my branch with the update I'd like to get confirmation that this resolves the error https://github.com/microsoft/coe-alm-accelerator-templates/blob/mochs-aa4pp-1793/Pipelines/Templates/update-deployment-settings.yml

Even if this does resolve the error there is still a potential issue with how steps and pipelines got out of sync to begin with e.g. why were there 0 steps or why was there potentially more pipelines than steps configured...

Any input on this last point would help.

**Comment by Ballzer0** [comment]
It is producing a new error now
```
2022-01-21T09:16:14.9272246Z ##[section]Starting: Update Deployment Settings
2022-01-21T09:16:14.9447229Z ==============================================================================
2022-01-21T09:16:14.9447596Z Task         : PowerShell
2022-01-21T09:16:14.9447893Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2022-01-21T09:16:14.9448170Z Version      : 2.194.0
2022-01-21T09:16:14.9448400Z Author       : Microsoft Corporation
2022-01-21T09:16:14.9448743Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2022-01-21T09:16:14.9449139Z ==============================================================================
2022-01-21T09:16:16.7209627Z Generating script.
2022-01-21T09:16:16.7679142Z ========================== Starting Command Output ===========================
2022-01-21T09:16:16.8031467Z ##[command]"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a\_temp\52a7d421-9246-4430-9c29-71798f195cb8.ps1'"
2022-01-21T09:16:17.1335554Z 
2022-01-21T09:16:17.1336088Z 
2022-01-21T09:16:17.1341605Z     Directory: D:\a\1\s\quantus-solutions\shi_core
2022-01-21T09:16:17.1342005Z 
2022-01-21T09:16:17.1342127Z 
2022-01-21T09:16:17.1357259Z Mode                LastWriteTime         Length Name                                                                  
2022-01-21T09:16:17.1367115Z ----                -------------         ------ ----                                                                  
2022-01-21T09:16:17.1381138Z d-----        1/21/2022   9:16 AM                config                                                                
2022-01-21T09:16:19.5361518Z 0
2022-01-21T09:16:19.5362135Z 1
2022-01-21T09:16:19.5382325Z 0
2022-01-21T09:16:20.3726033Z ##[warning]Invalid JSON was supplied for parameter configurationData.
2022-01-21T09:16:21.2579405Z Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
2022-01-21T09:16:21.2580485Z At D:\a\_temp\52a7d421-9246-4430-9c29-71798f195cb8.ps1:264 char:48
2022-01-21T09:16:21.2581293Z + ... Variables | Add-Member -MemberType NoteProperty -Name $configurationV ...
2022-01-21T09:16:21.2581839Z +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2022-01-21T09:16:21.2582595Z     + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
2022-01-21T09:16:21.2583203Z     + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo 
2022-01-21T09:16:21.2583831Z    mmand
2022-01-21T09:16:21.2584205Z  
2022-01-21T09:16:21.3558937Z ##[error]PowerShell exited with code '1'.
2022-01-21T09:16:21.3966020Z ##[section]Finishing: Update Deployment Settings
```

**Comment by Ballzer0** [comment]
I have updated from the preview to the **ALM accelerator for Power Platform** instead now, and it seems to work when i use the code [https://github.com/microsoft/coe-alm-accelerator-templates/blob/mochs-aa4pp-1793/Pipelines/Templates/update-deployment-settings.yml](url) you provided.

**Comment by mdpc-nnit** [comment]
MIne seems to work now, however, I encounter the following warning a lot, which results in no deploy variables get updated.

Also, is it supposed to create new pipelines in azure devops now? Because I haven't got it to work, and since we are using a specific folder structure (that is, pipeline folders, not folders in the repo, where I think we follow your guidance), I am a bit worried that it will not fit with our setup.

**Comment by mikefactorial** [comment]
@mdpc-nnit as far as creating the pipelines the build service in AzDO will need additional permissions, mentioned in the release notes but not explicitly, that may not have been applied with the upgrade. Take a look here to check if the additional permissions were added. https://docs.microsoft.com/en-us/power-platform/guidance/coe/setup-almacceleratorpowerplatform#update-permissions-for-the-project-build-service

**Comment by mikefactorial** [comment]
@mdpc-nnit for the warning message can you verify the variables being passed into the export pipeline? In particular you are seeing a value in the following variables.

![image](https://user-images.githubusercontent.com/42348035/150534417-508b79b8-4de0-4b7c-a47f-8bf7cd08477e.png)

**Comment by mdpc-nnit** [comment]
`EnvironmentName` is undefined. Data has a value, but in one case, we have
`[{"EnvironmentUrl":"https://xxxx.crm4.dynamics.com/""UserSettingId":[]}]`
Does that look rigt?

**Comment by mdpc-nnit** [comment]
Did create a separate issue for this: https://github.com/microsoft/coe-starter-kit/issues/1818

**Comment by Ballzer0** [comment]
> `EnvironmentName` is undefined. Data has a value, but in one case, we have `[{"EnvironmentUrl":"https://xxxx.crm4.dynamics.com/""UserSettingId":[]}]` Does that look right?

i also had that issue. i had to manually update the environment variable of the generated pipeline, something seems a bit off here

**Comment by mikefactorial** [comment]
Closing this issue will respond in #1818

**Comment by mikefactorial** [comment]
@camurao have you looked at issue #1827 for the resolution? This is a permissions issue.

**Comment by camurao** [comment]
@mikefactorial Thanks Mike. After setting up the rest of the security components, it finally worked for me!

**Comment by CoEStarterKitBot** [comment]
@mdpc-nnit This has been fixed in the latest release. Please install [the latest version of the toolkit](https://aka.ms/CoEStarterKitDownload) following the instructions for [installing updates](https://docs.microsoft.com/power-platform/guidance/coe/setup#installing-updates). Note that if you do not remove the unmanaged layers as described there you will not receive updates from us.

**Comment by bhatiarahul1993** [comment]
Hi, 
I am trying ALM accelerator and  I am unable to commit the  solution to git repo. Below is screenshot of error message. 

![image](https://user-images.githubusercontent.com/35290041/177581026-c0e5475a-5ff6-4532-905a-ec199e3e2515.png)
